c10/test/core/DispatchKeySet_test.cpp

```diff
@@ -53,3 +53,53 @@ TEST(DispatchKeySet, Full) {
     ASSERT_TRUE(full.has(tid));
   }
 }
+
+TEST(DispatchKeySet, IteratorBasicOps) {
+  DispatchKeySet empty_set;
+  DispatchKeySet full_set(DispatchKeySet::FULL);
+  DispatchKeySet mutated_set = empty_set.add(static_cast<DispatchKey>(1));
+
+  // Constructor + Comparison
+  ASSERT_EQ(*empty_set.begin(), DispatchKey::NumDispatchKeys);
+  ASSERT_EQ(*empty_set.end(), DispatchKey::NumDispatchKeys);
+  ASSERT_EQ(*mutated_set.begin(), static_cast<DispatchKey>(1));
+
+  ASSERT_TRUE(empty_set.begin() == empty_set.end());
+  ASSERT_TRUE(full_set.begin() != full_set.end());
+
+  // Increment Ops
+  ASSERT_TRUE(full_set.begin() == full_set.begin()++);
+  ASSERT_TRUE(full_set.begin() != ++full_set.begin());
+}
+
+TEST(DispatchKeySet, IteratorEmpty) {
+  DispatchKeySet empty_set;
+  uint8_t i = 0;
+
+  for (auto it = empty_set.begin(); it != empty_set.end(); ++it) {
+    i++;
+  }
+  ASSERT_EQ(i, 0);
+}
+
+TEST(DispatchKeySet, IteratorFull) {
+  DispatchKeySet full_set(DispatchKeySet::FULL);
+  uint8_t i = 0;
+
+  for (auto it = full_set.begin(); it != full_set.end(); ++it) {
+    i++;
+  }
+  ASSERT_EQ(i, static_cast<uint8_t>(DispatchKey::NumDispatchKeys) - 1);
```

Comment:
This looks pretty odd. If there are three dispatch keys, if I iterate through each of them, incrementing i once per, then I'd expect i == 3! (Not 2!) Maybe this is right, but it smells like an off-by-one error to me.

Reply:
I don't believe it is. `NumDispatchKeys` is an actual DispatchKey enum (and has a uint8_t conversion) that always comes last. Additionally, the DispatchKey that has a value of zero, represents undefined or an empty key set. Consequentially, If there are three dispatch keys, `NumDispatchKeys` will have a unit8_t value of 4.

Reply:
You're right, zero is not included, so -1.